content/blog/2021-12-07-ml-experiment-versioning.md

```diff
@@ -0,0 +1,219 @@
+this meta-information across lots of experiments.
+
+Git can't manage or compare all that experiment meta-information, but it is
+still better for code. Tools like GitHub make distributed collaboration easy,
+and you can see incremental code changes. That's why experiments get split
+between Git for code and experiment tracking tools for meta-information (usually
+with a link in one or the other to keep track).
+
+ML experiment versioning combines experiment tracking and version control.
+Instead of managing these separately, keep everything in one place and get the
+benefits of both, like:
+
+- **Experiments as code**: Track meta-information in the repository and version
+  it like code.
+- **Versioned reproducibility**: Save and restore experiment state, and track
+  changes to only execute what's new.
+- **Distributed experiments**: Organize locally and choose what to share,
+  reusing your existing repo setup.
+
+![Experiment Versioning](2021-12-07/exp-versioning.png)
+
+# ML Experiments as Code
+
+Experiment versioning treats experiments as code. It saves all metrics,
+hyperparameters, and artifact information in text files that can be versioned by
+Git (DVC [data versioning](https://dvc.org/doc/start/data-and-model-versioning)
+backs up the artifacts themselves anywhere). You do not need a centralized
+database or online services. Git becomes a store for experiment
+meta-information.
+
+You can choose your own file formats and paths, which you can configure in DVC:
+
+```dvc
+$ dvc exp init -i
+This command will guide you to set up a default stage in dvc.yaml.
+See https://dvc.org/doc/user-guide/project-structure/pipelines-files.
+
+DVC assumes the following workspace structure:
+├── data
+├── metrics.json
+├── models
+├── params.yaml
+├── plots
+└── src
+
+Command to execute: python src/train.py
+Path to a code file/directory [src, n to omit]: src/train.py
+Path to a data file/directory [data, n to omit]: data/images/
+Path to a model file/directory [models, n to omit]:
+Path to a parameters file [params.yaml, n to omit]:
+Path to a metrics file [metrics.json, n to omit]:
+Path to a plots file/directory [plots, n to omit]: logs.csv
+```
+
+Once you set up your repo in this structure, you start to see the benefits of
+this approach. Experiment meta-information lives in readable files that are
+always available, and your code can stay clean. You can read, save, and version
+your meta-information:
+
+```dvc
+$ cat params.yaml
+train:
+  epochs: 10
+model:
+  conv_units: 128
+```
+
+```dvc
+$ cat metrics.json
+{"loss": 0.24310708045959473, "acc": 0.9182999730110168}
+```
+
+You can see what changed in parameters, code, or anything else:
+
+```dvc
+$ git diff HEAD~1 -- params.yaml
+diff --git a/params.yaml b/params.yaml
+index baad571a2..57d098495 100644
+--- a/params.yaml
++++ b/params.yaml
+@@ -1,5 +1,5 @@
+ train:
+   epochs: 10
+-model:
+-  conv_units: 16
++model:
++  conv_units: 128
+```
```

Comment:
When I did Batuhan's post I learned how to do this. It shows the diffs better:
![image](https://user-images.githubusercontent.com/34289199/144770091-fff6e02a-511f-48ad-8a8f-9fbdebc52f4a.png) 
![image](https://user-images.githubusercontent.com/34289199/144770138-7bd215e3-9c50-4978-abaf-c870387faba7.png)

Would that be better?  or does it need to be in DVC?